### DISCUSSION TITLE
`TS_RS_EXPORT_DIR` quirks

### DISCUSSION THREAD

**Discussion by escritorio-gustavo** [Aleph-Alpha/ts-rs#245] [General]
Hey @NyxCode! There are a few things related to the `TS_RS_EXPORT_DIR` environment variable and `#[ts(export_to = "...")]` that I'd like to discuss. Almost all of which are breaking changes:

- [x] We need a way to clean paths containing `..` as that can cause problems in `diff_paths` (done in #247)
- [x] When we use `TS_RS_EXPORT_DIR` and `#[ts(export_to = "...")]` the path in `#[ts(export_to = "...")]` is still relative to `Cargo.toml`, and not to `TS_RS_EXPORT_DIR`. Should we change this? (done in #250 + #256)
- [ ] We determine if the path given to `#[ts(export_to = "...")]` is a directory by whether or not it ends in a slash `/`. I think we should change this to detect a `.ts` file extension instead because:
    - It's way too easy to forget
    - Not all operating systems use `/` as the directory separator (Windows accepts it, but it actually uses the backslash `\`)

**Comment by NyxCode** [comment]
> When we use TS_RS_EXPORT_DIR and #[ts(export_to = "...")] the path in #[ts(export_to = "...")] is still relative to Cargo.toml, and not to TS_RS_EXPORT_DIR. Should we change this? 

Hm, you're right! That's very unintuitive! I think we should definetely change this!  
Since we haven't published a release with `TS_RS_EXPORT_DIR` yet, now is the best time to fix this.

> We determine if the path given to #[ts(export_to = "...")] is a directory by whether or not it ends in a slash /. I think we should change this to detect a file extension

This tripped me up a couple of times as well. I do think the current behaviour is "correct" and how many tools handle this, but it's a bit unintuitive, since exporting to a file without extension is most likely not what any user wants, ever. 
 
If we changed this, it would become _impossible_ to export to a file without extension though. Probably not a (common) usecase, but it's still something which would become completely impossible.  
So yeah, on this, I don't have a strong opinion. If I had to decide, then I'd keep it as-is, but I wouldn't be against changing it either.

> We need a way to clean paths containing .. as that can cause problems in diff_paths
We could use the path-clean crate for this

Do you have an example for this?

**Reply by escritorio-gustavo** [reply]
> Do you have an example for this?

I do remember @stegaBOB mentioning issues with it back in #195. I'll check if I can reproduce it

**Reply by escritorio-gustavo** [reply]
> If we changed this, it would become impossible to export to a file without extension though. Probably not a (common) usecase, but it's still something which would become completely impossible.

We could make a trailing `.` mean a file without an extension, since no file can have a trailing dot, this should be a hacky, but working solution

I don't see why anyone would ever want a file without the `.ts` extension though, since that would make it very hard for `tsc` to pick up and type check that file

> If I had to decide, then I'd keep it as-is

That's fine with me too, having to use the `/` at the end is not that big a deal. I do think we should also support `\` though

**Reply by escritorio-gustavo** [reply]
> > Do you have an example for this?
> 
> I do remember @stegaBOB mentioning issues with it back in #195. I'll check if I can reproduce it


Got it! Check out #247

**Comment by NyxCode** [comment]
> Got it! Check out https://github.com/Aleph-Alpha/ts-rs/pull/247

Ah, thanks for reproducing it!  
If, for example, `base` is `../` and `path` is `a/`, then `diff_paths` has no clue how to get to `a/` from one level up.  

However: If we started with an absolute path (by e.g doing `var("CARGO_MANIFEST_DIR")?.join(path)`), then we'd get something like `/home/a/b/c/../`. Here, it'd be possible to get rid of the `../` component, since it would cancel itself out with the `c/` component before it.  

There's [std::path::absolute](https://doc.rust-lang.org/std/path/fn.absolute.html) on nightly which does that, but I think we can implement a crappy version ourself.

**Reply by escritorio-gustavo** [reply]
> I think we can implement a crappy version ourself.

I just pushed a commit to do that. However, I think we should change `diff_paths` to return `Result`, because to keep the `Option` I had to ignore an error

**Reply by NyxCode** [reply]
Noooooooo, you were faster than me, again! 😆

**Reply by escritorio-gustavo** [reply]
> There's [std::path::absolute](https://doc.rust-lang.org/std/path/fn.absolute.html) on nightly

Man, it feels like everything we need to make our lives easier is in nightly lol

**Reply by NyxCode** [reply]
welcome to rust 😂

**Comment by NyxCode** [comment]
Here's what I meant:
```rs
// paths we start with
let base = "../";
let path = "a/";
println!("from {base:?} to {path:?}");

// make both absolute (so that they start at the root). "." and ".." will remain, however
let manifest_dir = std::env::var("CARGO_MANIFEST_DIR").unwrap();
let manifest_dir = Path::new(&manifest_dir);
let base = manifest_dir.join(Component::ParentDir);
let path = manifest_dir.join("a");
println!("from {base:?} to {path:?}");

// remove all ".." or "." from both paths
let base = make_absolute(base);
let path = make_absolute(path);
println!("from {base:?} to {path:?}");

// remove the prefix both paths have in common
let path: PathBuf = {
    let mut path = path.components().collect::<VecDeque<_>>();
    let mut base = base.components().collect::<VecDeque<_>>();
    while path.front().is_some() && path.front() == base.front() {
        path.pop_front();
        base.pop_front();
    }
    path.into_iter().collect()
};
println!("=> {:?}", path);
```

Here, we turn 
`../` and `a/`
into
`/home/some/path/../` and `/home/some/path/a/`
into 
`/home/some/` and `/home/some/path/a`
into
` ` and `path/a/`

The function `make_absolute` removes all `".."` and `"."` from the paths, and looks like this:
```rs
fn make_absolute(p: impl AsRef<Path>) -> PathBuf {
    let p = p.as_ref();
    assert!(p.is_absolute(), "path must start with C:/ or /");

    let mut out = Vec::new();
    for c in p.components() {
        match c {
            Component::CurDir => (),
            Component::ParentDir => {
                out.pop().expect("path is referring to something above root");
            },
            other => {
                out.push(other);
            },
        }
    }

    out.into_iter().collect()
}
```

I'm sure there's a lot of room for improvement, but it's a start, I think.

**Reply by NyxCode** [reply]
Hm, seems like I re-implemented `path_clean`, lol

**Reply by escritorio-gustavo** [reply]
I think it's a better idea to use `std::env::current_dir` than `CARGO_MANIFEST_DIR`, as that's how it is currently implemented in `std::path::absolute`

**Reply by escritorio-gustavo** [reply]
Copied from https://doc.rust-lang.org/src/std/sys/unix/path.rs.html
```rs
/// Make a POSIX path absolute without changing its semantics.
pub(crate) fn absolute(path: &Path) -> io::Result<PathBuf> {
    // This is mostly a wrapper around collecting `Path::components`, with
    // exceptions made where this conflicts with the POSIX specification.
    // See 4.13 Pathname Resolution, IEEE Std 1003.1-2017
    // https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13

    // Get the components, skipping the redundant leading "." component if it exists.
    let mut components = path.strip_prefix(".").unwrap_or(path).components();
    let path_os = path.as_os_str().as_encoded_bytes();

    let mut normalized = if path.is_absolute() {
        // "If a pathname begins with two successive <slash> characters, the
        // first component following the leading <slash> characters may be
        // interpreted in an implementation-defined manner, although more than
        // two leading <slash> characters shall be treated as a single <slash>
        // character."
        if path_os.starts_with(b"//") && !path_os.starts_with(b"///") {
            components.next();
            PathBuf::from("//")
        } else {
            PathBuf::new()
        }
    } else {
        env::current_dir()?
    };
    normalized.extend(components);

    // "Interfaces using pathname resolution may specify additional constraints
    // when a pathname that does not name an existing directory contains at
    // least one non- <slash> character and contains one or more trailing
    // <slash> characters".
    // A trailing <slash> is also meaningful if "a symbolic link is
    // encountered during pathname resolution".
    if path_os.ends_with(b"/") {
        normalized.push("");
    }

    Ok(normalized)
}
```

**Reply by NyxCode** [reply]
> I think it's a better idea to use std::env::current_dir than CARGO_MANIFEST_DIR, as that's how it is currently implemented in std::path::absolute

Hmm. I felt like `CARGO_MANIFEST_DIR` makes more sense, since that's what the paths are actually supposed to be relative to. But as far as I can tell, `CARGO_MANIFEST_DIR == current_dir()` when running `cargo test`, no matter what I try. So maybe it doesn't matter in this use-case?

**Reply by escritorio-gustavo** [reply]
> But as far as I can tell, `CARGO_MANIFEST_DIR == current_dir()`

Looks that way to me too. Maybe the test binary is always executed from `CARGO_MANIFEST_DIR`?

**Reply by escritorio-gustavo** [reply]
Found a difference! In #248 `CARGO_MANIFEST_DIR` causes CI to fail because the types in the `example` directory go to `ts-rs/bindings` instead of `example/bindings`

**Comment by escritorio-gustavo** [comment]
> If we changed this, it would become _impossible_ to export to a file without extension though. Probably not a (common) usecase, but it's still something which would become completely impossible.

We saw in [this test run](https://github.com/Aleph-Alpha/ts-rs/actions/runs/8616297488/job/23613599284) (which was fixed by https://github.com/Aleph-Alpha/ts-rs/pull/293/commits/6c3b0cb4dbd7434957c2ed027ef0fa55cf001810) that files without extensions already break exporting when the `format` feature is enabled, so is this really a problem?
